serialization/serialization.go

```diff
@@ -0,0 +1,283 @@
+					f.Value = uintval
+				} else {
+					return errors.New("unexpected type, expecting uint64")
+				}
+				m.Fields[i].Type = f
+			case FieldIntVar:
+				var val interface{}
+				val, n, err = decodeVar(data, pos, false)
+				if err != nil {
+					return err
+				}
+				if intval, ok := val.(int64); ok {
+					f.Value = intval
+				} else {
+					return errors.New("unexpected type, expecting int64")
+				}
+				m.Fields[i].Type = f
+			case FieldString:
+				f.Value, n, err = decodeString(data, pos)
+				if err != nil {
+					return err
+				}
+				m.Fields[i].Type = f
+			default:
+				return fmt.Errorf("unsupported field type: %T", m.Fields[i].Type)
+			}
+			pos = n
+		}
+
+	default:
+		return fmt.Errorf("unsupported type: %T", v)
+	}
+	return nil
+}
+
+func decodeString(data []byte, pos uint64) (string, uint64, error) {
+	if len(data) < int(pos)+1 {
+		return "", pos, errors.New("string truncated, expected at least one byte")
+	}
+	strLen := int(data[pos] >> 1)
+	pos++
+	if len(data) < int(pos)+strLen {
+		return "", pos, fmt.Errorf("string truncated, expected length: %d", strLen)
+	}
+	return string(data[pos : pos+uint64(strLen)]), pos + uint64(strLen), nil
+}
+
+func decodeFixed(data []byte, pos uint64, intlen int) ([]byte, uint64, error) {
+	var b bytes.Buffer
+
+	for {
+		if len(data) < int(pos)+1 {
+			return b.Bytes(), pos, errors.New("data truncated")
+		}
+		if data[pos]%2 == 0 {
+			b.WriteByte(data[pos] >> 1)
+		} else {
+			if len(data) < int(pos)+2 {
+				return b.Bytes(), pos, errors.New("data truncated")
+			}
+			switch data[pos+1] {
+			case 0x2:
+				b.WriteByte((data[pos] >> 2) + 0x80)
+			case 0x3:
+				b.WriteByte((data[pos] >> 2) + 0xc0)
+			default:
+				return nil, pos, fmt.Errorf("unknown decoding for %v", data[pos])
+			}
+			pos++
+		}
+		pos++
+		if b.Len() == intlen {
+			break
+		}
+	}
+	return b.Bytes(), pos, nil
+}
+
+func decodeVar(data []byte, pos uint64, unsigned bool) (interface{}, uint64, error) {
+	if len(data) < int(pos)+1 {
+		return 0, pos, errors.New("data truncated")
+	}
+	flen := trailingOneBitCount(data[pos]) + 1
+	if len(data) < int(pos)+flen {
+		return 0, pos, fmt.Errorf("truncated data, expected length: %d", flen)
+	}
+	var tNum uint64
+	switch flen {
+	case 1:
+		tNum = uint64(data[pos])
+	case 2:
+		tNum = uint64(binary.LittleEndian.Uint16(data[pos : int(pos)+flen]))
+	case 3:
+		tNum = uint64(binary.LittleEndian.Uint32(
+			slices.Concat(data[pos:int(pos)+flen], []byte{0x0})))
+	case 4:
+		tNum = uint64(binary.LittleEndian.Uint32(data[pos : int(pos)+flen]))
+	case 5:
+		tNum = binary.LittleEndian.Uint64(
+			slices.Concat(data[pos:int(pos)+flen], []byte{0x0, 0x0, 0x0}))
+	case 6:
+		tNum = binary.LittleEndian.Uint64(
+			slices.Concat(data[pos:int(pos)+flen], []byte{0x0, 0x0}))
+	case 7:
+		tNum = binary.LittleEndian.Uint64(
+			slices.Concat(data[pos:int(pos)+flen], []byte{0x0}))
+	case 8:
+		tNum = binary.LittleEndian.Uint64(data[pos : int(pos)+flen])
+	}
```

Comment:
simpler would be to have an `[8]byte bytes` which you `copy(bytes, data[pos:int(pos)+flen])` & then you can skip switch with `tNum := binary.LittleEndian.Uint64(bytes[:])`
```suggestion
	var bytes [8]byte
	copy(bytes[:], data[pos:int(pos)+flen])
	tNum := binary.LittleEndian.Uint64(bytes[:])
```